src/CloudNative.CloudEvents/Core/BinaryDataUtilities.cs

```diff
@@ -41,5 +44,59 @@ public static byte[] ToByteArray(Stream stream)
             stream.CopyTo(memory);
             return memory.ToArray();
         }
+
+        /// <summary>
+        /// Returns a read-only <see cref="MemoryStream"/> view over the given memory.
+        /// </summary>
+        /// <param name="memory">The memory to create a stream view over.</param>
+        /// <returns>A read-only stream view over <paramref name="memory"/>.</returns>
+        public static MemoryStream AsStream(ReadOnlyMemory<byte> memory)
+        {
+            var segment = GetArraySegment(memory);
+            return new MemoryStream(segment.Array, segment.Offset, segment.Count, false);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="memory"></param>
+        /// <param name="encoding"></param>
+        public static string GetString(ReadOnlyMemory<byte> memory, Encoding encoding)
+        {
+            // TODO: Use encoding.GetString(memory.Span) where we can?
+            var segment = GetArraySegment(memory);
+            return encoding.GetString(segment.Array, segment.Offset, segment.Count);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="source"></param>
+        /// <param name="destination"></param>
+        /// <returns></returns>
+        public static async Task CopyToStreamAsync(ReadOnlyMemory<byte> source, Stream destination)
+        {
+            var segment = GetArraySegment(source);
+            await destination.WriteAsync(segment.Array, segment.Offset, segment.Count).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// TODO: Basically this is a sometimes-more-efficient version of ToArray
+        /// </summary>
+        /// <param name="memory"></param>
+        /// <returns></returns>
+        public static byte[] ToArray(ReadOnlyMemory<byte> memory) =>
+            !MemoryMarshal.TryGetArray(memory, out var segment) || segment.Offset == 0 || segment.Count != memory.Length
+            ? memory.ToArray()
+            : segment.Array;
+
+        private static ArraySegment<byte> GetArraySegment(ReadOnlyMemory<byte> memory)
+        {
+            if (!MemoryMarshal.TryGetArray(memory, out var segment))
+            {
+                throw new ArgumentException("Unable to obtain array from memory");
```

Comment:
Interesting.. why do we throw here as opposed to falling back to ToArray?

Reply:
I'd been hoping to avoid copying as far as possible, and requiring that input data *could* be obtained as an array. We should chat about this.